File: Medium/merge_intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if(intervals.empty())
            return {{}};
        int n = intervals.size();
        sort(intervals.begin(), intervals.end(), [&] (const auto &x, const auto &y) {
            return x < y;
        });
        vector<vector<int>> ret;
        int start = intervals[0][0];
        int end = intervals[0][1];
        ret.push_back(vector<int>{start, end});
        for(int i = 1; i < n; i++) {
            if (end >= intervals[i][0]) {
                if (end < intervals[i][1]) {
                    ret.back().pop_back();
                    ret.back().push_back(intervals[i][1]);
                    end = intervals[i][1];
                }
            } else {
                start = intervals[i][0];
                end = intervals[i][1];
                ret.push_back(vector<int> {start, end});
            }
        }
        return ret;
    }
};

File: 56-merge-intervals/merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        int n = intervals.length;
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        List<int[]> list = new ArrayList<>();
        int l = 0;
        int r = 1;
        list.add(new int[]{intervals[l][0], intervals[l][1]});
        while (r < n) {
            if (list.get(list.size()-1)[1] >= intervals[r][0]) {
                list.set(list.size() - 1, new int[]{intervals[l][0], Math.max(list.get(list.size()-1)[1], intervals[r][1])});
                r += 1;
            } else {
                list.add(new int[]{intervals[r][0], intervals[r][1]});
                l = r;
                r++;
            }
        }
       
        int[][] answer = new int[list.size()][2];
        int i = 0;
        for (int[] interval: list) {
            answer[i] = interval;
            i++;
        }
        return answer;
    }
}